### ISSUE TITLE
Should we separate the capturing groups from the subpatterns?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#10]
It is still not evident, if we need to bind the capturing groups to subpatterns, or they should be something separate.

**Comment by inf3rno** [comment]
I'd like to do compilation and adding capturing groups totally in the background. So the regex pattern and the groups should depend on the expectations about the results. I am not sure how to do that. Let's start with our classic example.

```js
var n = number();
var w = word();
var x = alternatives(n,w);
var p = quantify(x,2);
var r = p.match("abc123");
{
	value: "abc123",
	pattern: p,
	subpatterns: [
		{
			value: "abc",
			pattern: x,
			subpatterns: [
				{
					value: "abc",
					pattern: w
				}
			]
		},
		{
			value: "123",
			pattern: x,
			subpatterns: [
				{
					value: "123",
					pattern: n
				}
			]
		}
	]
}
```

**Comment by inf3rno** [comment]
The subpatterns must be reusable here, but the capturing must be configurable. How do we do that?
For example we have `p1,p2,p3` where `p1` captures only `n`-s, `p2` captures only `w`-s and `p3` captures both. How do we solve that?

**Comment by inf3rno** [comment]
The first approach is separating capturing groups from subpatterns and add them to the definition:
```js
var n = capture(number());
var w = word();
var x = alternatives(n,w);
var p1 = quantify(x,2);

var n = number();
var w = capture(word());
var x = alternatives(n,w);
var p2 = quantify(x,2);

var n = number();
var w = word();
var x = alternatives(n,w);
var p3 = concatenate(x, capture(x));
```

**Comment by inf3rno** [comment]
With this approach we cannot reuse `n1`, `w2`, `p3` and since `n` and `w` are dependencies, we cannot reuse `x1`, `x2`, `p1`, `p2` either. I think this is unacceptable, since in most of the cases we will have dependency trees like this.

**Comment by inf3rno** [comment]
A possible workaround is copying and clearing capturing groups:
```js
var n1 = capture(number());
var w = word();
var x1 = alternatives(n1,w);
var p1 = quantify(x,2);
```
```js
var p2 =clear(deepCopy(p1));
replace(p2, find(p2,w), capture(find(p2, w)));
```

**Comment by inf3rno** [comment]
It is far from evident in the case of `p3`, because we need to find somehow the second index on the quantifier. Ofc. for other approaches that is not evident too, we need to talk about this later and for now exclude that example and work only on the first 2.

**Comment by inf3rno** [comment]
I don't think that editing the subpattern hierarchy is a good solution for reusing subpatterns. It looks weird and not that easy to follow.

**Comment by inf3rno** [comment]
Another solution is having a separate compilation function which contains the info for the capturing groups.

```js
var n = number();
var w = word();
var x = alternatives(n,w);
var p = quantify(x,2);
```

```js
var p1 = compile(p).query(n);
var p2 = compile(p).query(w);
```

We could improve this by applying lazy compile and do only the query part on the fly.

```js
var r1 = p.query(n).match("abc123");
var r2 = p.query(w).match("abc123");
```
On the fly compilation is not that good though. We need to keep the subpatterns in memory for it. Better to do compilation and query creation first and let the GC remove the subpatterns if we no longer need it.

**Comment by inf3rno** [comment]
Actually we do the same by the previous approach too. That is basicly copy + clear + query + replace. So this querying is definitely better than copying and it is a lot easier to reuse the patterns with it. The only drawback that we have to write long pathes with it sometimes. E.g. in the case of `p3` we could write:

```js
var p3 = compile(p).query(path(x,1,w)).query(path(x,1,n));
```

**Comment by inf3rno** [comment]
What makes querying a little bit inconvenient that we cannot add a query while we are building the pattern. So these are 2 different steps. But we could do something like this for example:

```js
var q1 = query(), q2 = query(), q3 = query();
var n = number();
q1.add(n);
var w = word();
q2.add(w);
var x = alternatives(n,w);
q3.add(n,w);
var p = quantify(x,2);
q3.constraint(hasParent([n,w], p.index(1)));
var p1 = compile(p).query(q1);
var p2 = compile(p).query(q2);
var p3 = compile(p).query(q3);
```

It does not look that ordered, but it might be easier to query this way. Another option to do something like
```js
var p1 = pattern(), p2 = pattern(), p3 = pattern();
var n = number();
p1.capture(n);
var w = word();
p2.capture(w);
var x = alternatives(n,w);
p3.capture(n,w);
var p = quantify(x,2);
p3.constraint(hasParent([n,w], p.index(1)));
p1.compile(p);
p2.compile(p);
p3.compile(p);
```
We can do lazy compile by doing `p1.define(p)` or `p1.set(p)` or something like that. Once the pattern is compiled I think it should not be changed and we could clear the subpattern objects from it and keep only their ids. But I am not entirely sure about that. I mean we still need to process the results based on some sort of ids. If not based on the subpattern objects, then we will need strings or store their id, or something like that. Adding conversion rules is a possible solution for that. But that is a different thread.

**Comment by inf3rno** [comment]
```js
p3.capture(n,w);
p3.constraint(hasParent([n,w], p.index(1)));
```
can be written simpler this way:
```js
p3.capture(n,w);
p3.constraint(n,w).hasParent(p.index(1));
```

**Comment by inf3rno** [comment]
The real question here that how should we separate them, because we should if we want to support reusing the same subpattern instead of copying it. Copy does not work because if we have a range quantifier instead of an exact number, then we won't know how many times we have to copy the actual thing.

**Comment by inf3rno** [comment]
What we made clear here that the capturing groups should not be part of the pattern definition, because that hinders the reusing of these patterns. We can reuse them effectively in different contexts if we can add context specific capturing groups. In one context we need to capture one subpattern, in another context we need to capture a different one. What is interesting here, that unlike regex, we can descibe both cases with the same pattern and 2 different regex patterns will be generated based on the different capturing groups.

**Comment by inf3rno** [comment]
I find this good, but I am still not sure if we must use subpattern object instead of naming them. For example if we include an email subpattern to something, then it might be inconvenient to get the building parts of that subpattern and query what we want to use from those. E.g. we could ban email addresses based on host name. We can do it this way:

```js
var p = compile(patterns.common.email).query(patterns.common.email.hostname);
```

or we can do it this way:

```js
var p = compile(patterns.common.email).query("hostname");
```

We should decide if we want to separate query creation and result processing from these subpatterns/building parts or we want to save them somehow as objects. But ofc. that is a different topic. #11

**Comment by inf3rno** [comment]
What we decided here, that we will use `pattern.capture(subpattern)` to create a capturing group. I am not sure about the return value and if we should support multiple params. Actually this selection can create a lot of capturing groups at once instead of creating only a single one. Each places we used the actual subpattern there will be a capturing group added... That's why we need the constraints... Actually I meant conditions or selectors or something like that. I am not sure why I wrote constraints.

**Comment by inf3rno** [comment]
An issue with this approach is backreference #15. Normally those are bound to capturing groups. But in this case our capturing group selector covers multiple capturing groups. Somehow we need to ensure by backreference, that we are talking about a single capturing group. We might need to infer too in some cases e.g. a 0-1 quantifier might mean exclusion and the backreference would automatically choose the other usage. Ofc this is an advanced feature. For now we need to do it manually and maybe later try to automate selection and ease usage that way.

**Comment by inf3rno** [comment]
From out date example we know that something is useful if we can name it. E.g. date, year, month, day, etc. That means if we want to reuse a pattern we will usually end up exporting these useful subpatterns along with it. So we can capture later what we want to. E.g. 
```js
var dateFinder = use(date);
dateFinder.capture(date.year, parseInt);
dateFinder.compile(transformations.toObject({year: date.year}));
var match = dateFinder.match("blah 1912. dec. 20 blah");
console.log(match.value.year); // 1912
```

So I think we don't need to separate capturing groups from subpatterns. On the other hand capturing groups are bound to subpattern usage instead of the subpattern itself, so in theory we need a subpattern usage selector #29 instead of a subpattern to describe the capturing group. The selector should select only a single usage in theory. It is not a problem if the subpattern is used only once, but if there is a pattern repetition (not backreference rep) and for example we need to capture the first, second, third usages and name them, then we might have a problem. Copy was a solution for that, but I think we need something different to do it.

Probably instead of `p = concatenate(a,a,a)` we need to do

```js
a1 = use(a)
a2 = use(a)
a3 = use(a)
p = concatenate(a1,a2,a3)
p.capture(a2);
```

This way we don't need a complicated selector to decide which usage to choose:

```js
p = concatenate(a,a,a)
p.capture(a, usage(2))
```

Actually both are ok, if we don't reserve the `use` for base pattern usage before compile, but the real example here is quantifiers.

**Comment by inf3rno** [comment]
For example if we have `p = quantify(a, 0, 3)` and we want to capture the second usage if it exists, then what should we do? How would we solve this in regex?

**Comment by inf3rno** [comment]
It should translate to:
```js
/a(a)a*|a?/
```

**Comment by inf3rno** [comment]
Is this possible with reuse?

Well only if we follow the same logic as the regex has:

```js
a2 = reuse(a);
x = alternatives(concatenate(a, a2, quantify(a, 0, Infinity)), quantify(a, 0, Infinity));
```

Actually this is just the same as using `copy` but with a different name. And it is still ugly compared to the selectors.

**Comment by inf3rno** [comment]
Ok, let's choose the selectors. How do we use backreference for the a2 in the same x?

**Comment by inf3rno** [comment]
Interesting. So we say here that we want to capture the second usage of "a" and after that use it in backreference. If it does not exist, then the backreference will be empty, so in theory it will work with the following regex code:

`/(?:a(a)a*|a?)\1/`

So if we are talking about digits it will match 1 or 12 or 122 or 1232 or 12342, 123453, etc... I am not sure about the 12, I think the proper regex would be `/(?:a(a)a*|a{0,2})\1/` instead.

**Comment by inf3rno** [comment]
Actually what I wanted to do is adding something to a concat in a backreference that was not used that in the concat. So the upper example is not good, because it uses a quantifier not a concat. Let's do again the `concat(a,a,a)` example.

```js
x = concat(a,a,a,reference(select(a).from(x).usage(2));
```
As you can see the `x` does not have a value here, so `select` won't work. We need some sort of keyword for `this` here, but I am not sure how I can solve this. Probably

```js
x = concat(a,a,a,reference(select(a).usage(2));
```
So if `x` does not exist yet, then by omitting the `from()` part of the selection can be a sign that we are talking about the actual subpattern. If we have more complex definitions like

```js
x = concat(a,a,concat(a,a,a), quantify(a,1,4), reference(select(a).usage(3)));
```
Then it might be a problem.

Another issue here that the reference can come before the usage, which is funny.

```js
x = concat(a, reference(select(a).usage(3)), a, a);
```
what is very interesting here, that it is not clear which is the third usage, since we can count the backreference as an usage too.

I think we should do it that way, so it should be equivalent with

```js
x = concat(a, a, reference(select(a).usage(2)), a);
```
I am not sure if in theory we can have a reference where this logic does not work, but for now we can keep it in the description instead of enforcing the order. It will make the code harder to understand though.

Another problem here that we need to fill the context of the select later, so we will need a postproces loop if another capturing group is dependent on a reference like this. On the other hand this is unlikely unless we introduce parent, etc. selectors like `reference(select(a).usage(2).parent())`. In this case if we add the reference to the same concat we will have an infinite loop.

**Comment by inf3rno** [comment]
I got another example here. It is the date example with multiple alternative formats.

```js
alternatives(
    concat(year, separator, month, separator, day),
    concat(day, separator, month, separator, year)
)
```

How should we handle this?

- we might need to change the order by alt depending on the region, because the default changes by region
- if we want to check for many formats, we need to capture year, month, day in all of them and return a Date object, so we need a selector that can do it, which is not trivial. we need to bind a capturing group to multiple usages here, but we have a precondition, that it should be used only once in the alt.


```js
yyyymmdd = concat(year, separator, month, separator, day)
ddmmyyyy = concat(day, separator, month, separator, year)
date = alternatives(yyyymmdd , ddmmyyyy)
p = use(date)
p.capture(year)
p.returns(year) // compile
m = p.match("1912-03-19");
m2 = p.match("19-03-1912");
console.log(m.value, m2.value); // 1912, 1912
```

this should work. we need to be able to affect the order of alternatives too. it is not a problem here, but if we have yymmdd and ddmmyy, then it really matters what is the default order. in some cases it is which makes sense e.g. we don't have a 13th month, but in other cases both would make sense. I am not sure, maybe we need region specific pattern templates instead of mutating the pattern in those cases, but actually we don't need to change the order, we just need a different transformation, that replaces the day and the year with each other. it would be an ugly hax though especially if we have a number range in the day pattern and no range in the year pattern. so I don't think this different transformation would work, unless we use the same pattern for both instead of just partially overlapping patterns. so what we have left is multiple date patterns by region, where we hardcode the order into the pattern or mutating patterns #32 where we set the order by compilation with some sort of input. both might be good solutions, I am not sure which to support.

**Comment by inf3rno** [comment]
actually we could do the region specific pattern template on the fly
```js
order = [yymmdd, ddmmyy]
customDate = alternatives(order...)
p = use(customDate)
p.capture(year)
p.returns(year)
m = p.match("12-03-10")
console.log(m.value) // 12
m2 = p.match("22-03-10")
console.log(m2.value) // 10
```

**Comment by inf3rno** [comment]
I am not sure yet how I can add this n:m relationship to the API, but it would be cool. Probably using multiple selectors would help here:

```js
capture(year, month, day, function (year, month, day){...})
```

**Comment by inf3rno** [comment]
this makes the compilation transformation interesting too. we need to use the selections there

```js
dateGroup = p.capture(year, month, day, function (year, month, day){...})
p.returns(dateGroup)
```

**Comment by inf3rno** [comment]
actually I am not sure if I need to use selectors or groups by the compilation to make the value of the match. probably groups are better if we really want to make a variable for them. it is not a problem, I mean if we really have n:m relationship between the subpattern usages and the capturing groups in this lib, then the capturing groups might have very different names as the subpatterns. in many cases though there will be naming collision. e.g. if we capture only the year, then we will have a yearSubpattern and a yearCapturing group which uses a selector for that subpattern. in those cases we don't necessarily need to give a name to the capturing group, we could use the year subpattern as a capturing group identifier. it will make the compilation a little bit slower though. in these cases we have 1:1 relationship, and we use an implicit rule to have sugar syntax. so I think this should be added later as sugar syntax and the default lib should not contain it. which means we'll need some sort of configurable `identifyCapturingGroup` function that we can replace with the sugar syntax plugin and that will find us the capturing group for the given subpattern if we have an 1:1 relationship. not sure what to do by 1:n and maybe n:m relationships. just an example

```js
dateGroup = p.capture(year, month, day, function (year, month, day){...})
p.returns(year)
```

how should the lib interpret this? should it throw an exception?

**Comment by inf3rno** [comment]
I'll need a factsheet and example code where I can follow the progress here without developing real code.

**Comment by inf3rno** [comment]
I started one in the wiki, but I am not sure that is the right place for it. I'd better to write tests instead I guess. I am not sure how to handle this on a repo level. I mean if I commit tests without implementation, then those will fail on the main branch. Should I add some sort of skip flag for them? Should I add a separate branch for them? Should I commit only when all tests are green even on a dev branch? I have a contradition here, I need to discuss this with others first.

**Comment by inf3rno** [comment]
My conclusion that things can continue here in the issues and the results of the decisions must be added to the code as tests. They could be added as examples too, but it is important to have tests against the lib as well, and I don't like to work twice. The workflow will be somewhat slower this way for a single issue, but I won't have to reread all issues (even the closed ones) if I pause the project for a few months. The wiki does not seem to be the best place for it, beucase I don't want to maintain a long documentation before the lib becomes stable. I could write a short documentation there, but I want to bind it to issues and I want to version it too. Afaik. it is possible to version the wiki separately, but writing markdown in the IDE instead of writing it in the github editor is circumstantial.

**Comment by inf3rno** [comment]
Now what I need is processing the closed issues and write tests for them.

**Comment by inf3rno** [comment]
I already separated capturing groups from subpatterns a long time ago, I don't know why this is still open. I'll read the whole issue again and close it after that.

**Comment by inf3rno** [comment]
I think I'll merge this with #27.
